source/ShipEffectsShader.cpp

```diff
@@ -0,0 +1,351 @@
+		"    color *= sobellish(uv);\n"
+		// TODO: Move switch to uniform, add more patterns
+		"    int switchint = 0;\n"
+		"    switch(switchint)\n"
+		"    {\n"
+		"      case 0:\n"
+		"        color *= gridPattern(color.a, uv);\n"
+		"        break;\n"
+		"    }\n"
+		"  }\n"
+		"  else if(recentHitCount != 0)\n"
+		"  {\n"
+		"    color = sobellish(uv) * shieldColor * recentDamage[0] * 0.4;\n"
+		"  }\n"
+
+		"  finalColor = color;\n"
+		"}\n"
+		"\n";
+
+	shader = Shader(vertexCode, fragmentCode);
+	scaleI = shader.Uniform("scale");
+	frameI = shader.Uniform("frame");
+	frameCountI = shader.Uniform("frameCount");
+	positionI = shader.Uniform("position");
+	transformI = shader.Uniform("transform");
+	blurI = shader.Uniform("blur");
+	clipI = shader.Uniform("clip");
+
+	recentHitsI = shader.Uniform("recentHits");
+	recentDamageI = shader.Uniform("recentDamage");
+	recentHitsCountI = shader.Uniform("recentHitCount");
+	ratioI = shader.Uniform("ratio");
+	sizeI = shader.Uniform("size");
+
+	fastI = shader.Uniform("isFast");
+
+	shieldColorI = shader.Uniform("shieldColor");
+
+	glUseProgram(shader.Object());
+	glUniform1i(shader.Uniform("tex"), 0);
+	glUseProgram(0);
+
+	// Generate the vertex data for drawing sprites.
+	glGenVertexArrays(1, &vao);
+	glBindVertexArray(vao);
+
+	glGenBuffers(1, &vbo);
+	glBindBuffer(GL_ARRAY_BUFFER, vbo);
+
+	GLfloat vertexData[] = {
+		-.5f, -.5f,
+		-.5f,  .5f,
+		 .5f, -.5f,
+		 .5f,  .5f
+	};
+	glBufferData(GL_ARRAY_BUFFER, sizeof(vertexData), vertexData, GL_STATIC_DRAW);
+
+	glEnableVertexAttribArray(shader.Attrib("vert"));
+	glVertexAttribPointer(shader.Attrib("vert"), 2, GL_FLOAT, GL_FALSE, 2 * sizeof(GLfloat), nullptr);
+
+	// unbind the VBO and VAO
+	glBindBuffer(GL_ARRAY_BUFFER, 0);
+	glBindVertexArray(0);
+}
+
+
+
+void ShipEffectsShader::Draw(const Body* body, const Point& position, const vector<pair<Point, double>>* recentHits,
+	const float zoom, const float frame, const string &shieldColor)
+{
+	if(!body->GetSprite())
+		return;
+
+	Bind();
+	Add(Prepare(body, position, recentHits, zoom, frame, shieldColor));
+	Unbind();
+}
+
+
+
+void ShipEffectsShader::SetCenter(Point newCenter)
+{
+	center = newCenter;
+}
+
+
+
+ShipEffectsShader::EffectItem ShipEffectsShader::Prepare(const Body* body, const Point& position,
+	const vector<pair<Point, double>>* recentHits, const float zoom, const float frame, const string &shieldColor)
+{
+	if(!body->GetSprite())
+		return {};
+
+	EffectItem item;
+	item.texture = body->GetSprite()->Texture();
+	item.frame = frame;
+	item.frameCount = body->GetSprite()->Frames();
+	// Position.
+	item.position[0] = static_cast<float>(position.X() * zoom);
+	item.position[1] = static_cast<float>(position.Y() * zoom);
+
+	// Get unit vectors in the direction of the object's width and height.
+	double width = body->Width();
+	double height = body->Height();
+	Point unit = body->Facing().Unit();
+	Point uw = unit * width;
+	Point uh = unit * height;
+
+	item.size = body->Radius();
```

Comment:
```suggestion
	item.size = 80.;
```

While play testing I found this has undesirable visual effects on smaller ships but still kept nice visual effects on larger ships.  Setting this value to static made all ships look nice.

Maybe this should just be hard coded in the shader.